client/service/src/client/client.rs

```diff
@@ -622,25 +622,25 @@ where
 						None
 					},
 				};
-				// Ensure parent chain is finalized to maintain invariant that
-				// finality is called sequentially. This will also send finality
-				// notifications for top 250 newly finalized blocks.
-				if finalized && parent_exists {
-					self.apply_finality_with_block_hash(
-						operation,
-						parent_hash,
-						None,
-						info.best_hash,
-						make_notifications,
-					)?;
-				}
 
 				operation.op.update_cache(new_cache);
 				storage_changes
 			},
 			None => None,
 		};
 
+		// Ensure parent chain is finalized to maintain invariant that finality is called
+		// sequentially.
+		if finalized && parent_exists {
+			self.apply_finality_with_block_hash(
+				operation,
+				parent_hash,
+				None,
+				info.best_hash,
+				make_notifications,
```

Comment:
The at line 694 is the current block hash. The function is invoked using the `parent_hash`. The current hash is not included in this operation.

The current hash is not passed because `apply_finality_with_block_hash` assumes that the blocks to mark as finalized are already stored by the backend and are only marked as final

Snip from `apply_finality...`
```
			operation.op.mark_finalized(BlockId::Hash(finalize_new.hash), None)?;
```

The currently imported block is excluded by from this call because it is directly imported as final.

Snip from `execute_and_import_block`
```
		let leaf_state = if finalized {
			NewBlockState::Final
		} else if is_new_best {
			NewBlockState::Best
		} else {
			NewBlockState::Normal
		};
		...
		operation.op.set_block_data(
			import_headers.post().clone(),
			body,
			indexed_body,
			justifications,
			leaf_state,
		)?;
```

Reply:
This marks as finalized all the blocks up to the parent. The current block hash is not included here, it will be appended at  line 694.

Current block is not included in the call to `apply_finality_with_block_hash` because this function cannot be used for blocks that are not already stored in the backend.

The currently imported one is treated specially, it is directly imported as Final (row 652)

Reply:
Now I have read the comment above :facepalm: